src/java/voldemort/store/readonly/chunk/ChunkedFileSet.java

```diff
@@ -544,47 +552,69 @@ public int getChunkForKey(byte[] key) {
                     return valueBuffer.array();
                 }
                 case READONLY_V2: {
-
-                    // Buffer for 'numKeyValues', 'keySize' and 'valueSize'
+                    // Buffer for 'numKeyValues' (a short), 'keySize' (an int) and 'valueSize' (another int)
                     int headerSize = ByteUtils.SIZE_OF_SHORT + (2 * ByteUtils.SIZE_OF_INT);
-                    ByteBuffer sizeBuffer = ByteBuffer.allocate(headerSize);
-                    dataFile.read(sizeBuffer, valueLocation);
+                    int fileSize = getDataFileSize(chunk);
+                    ByteBuffer headerBuffer, keyBuffer, valueBuffer;
+                    headerBuffer = ByteBuffer.allocate(headerSize);
+                    dataFile.read(headerBuffer, valueLocation);
                     valueLocation += headerSize;
 
                     // Read the number of key-values
-                    short numKeyValues = sizeBuffer.getShort(0);
+                    short numKeyValues = headerBuffer.getShort(0);
 
                     // Read the key size
-                    int keySize = sizeBuffer.getInt(ByteUtils.SIZE_OF_SHORT);
+                    int keySize = headerBuffer.getInt(ByteUtils.SIZE_OF_SHORT);
 
                     // Read the value size
-                    int valueSize = sizeBuffer.getInt(ByteUtils.SIZE_OF_SHORT
-                                                      + ByteUtils.SIZE_OF_INT);
+                    int valueSize = headerBuffer.getInt(ByteUtils.SIZE_OF_SHORT + ByteUtils.SIZE_OF_INT);
 
                     do {
+                        if (valueLocation >= (fileSize - headerSize)) {
```

Comment:
@gnb, I added the check you requested for not going past the data file's length here.